tests/utilities/test_auto_restart.py

```diff
@@ -1060,3 +1061,137 @@ def run(should_fail, resume):
+        max_epochs=1,
+        limit_train_batches=4,
+        limit_val_batches=4,
+        val_check_interval=val_check_interval,
+        num_sanity_val_steps=0,
+    )
+
+    trainer = Trainer(**trainer_kwargs)
+    if should_signal:
+        with pytest.raises(ExitGracefullyException, match=status):
+            trainer.fit(model)
+    else:
+        trainer.fit(model)
+    assert trainer._terminate_gracefully == should_signal
+
+    return model
+
+
+@pytest.mark.parametrize("on_last_batch", [False, True])
+@pytest.mark.parametrize("val_check_interval", [0.5, 1.0])
+@pytest.mark.parametrize("failure_on_training", [False, True])
+@pytest.mark.parametrize("failure_on_step", [False, True])
+@mock.patch.dict(os.environ, {"PL_FAULT_TOLERANT_TRAINING": "1"})
+@RunIf(min_torch="1.7.0", skip_windows=True)
+def test_auto_restart_under_signal(on_last_batch, val_check_interval, failure_on_training, failure_on_step, tmpdir):
+    """This test asserts that if a signal is being sent during the training / validation phase, the model should
+    restart in a reproducible way."""
+
+    model_total = _fit_model(tmpdir, False, val_check_interval, failure_on_step, failure_on_training, on_last_batch)
+
+    if failure_on_step:
+        if on_last_batch:
+            if failure_on_training:
+                # Breaking on first validation batch.
+                # This is done to capture the random state of the validation dataloader.
+                status = "EvaluationEpochLoop:advance"
+            else:
+                # when breaking on last batch of validation, we should exist on `run_end` val_check_interval == 1.0
+                status = (
+                    "TrainingEpochLoop:on_run_end" if val_check_interval == 1.0 else "TrainingEpochLoop:on_advance_end"
+                )
+        else:
+            status = "TrainingEpochLoop:on_advance_end" if failure_on_training else "EvaluationEpochLoop:advance"
+    else:
+        if val_check_interval == 1.0:
+            status = "TrainingEpochLoop:on_run_end"
+        else:
+            # `training_epoch_end` happens after `validation_epoch_end` since Lightning v1.4
+            status = "TrainingEpochLoop:on_run_end" if failure_on_training else "TrainingEpochLoop:on_advance_end"
+
+    model_signaled = _fit_model(
+        tmpdir, True, val_check_interval, failure_on_step, failure_on_training, on_last_batch, status=status
+    )
+    checkpoint_path = str(tmpdir / ".pl_auto_save.ckpt")
+    assert os.path.exists(checkpoint_path)
+    model_restarted = _fit_model(tmpdir, False, val_check_interval, failure_on_step, failure_on_training, on_last_batch)
+
+    # check the batches
+    actual = torch.cat(model_signaled.seen_train_batches + model_restarted.seen_train_batches)
+    expected = torch.cat(model_total.seen_train_batches)
+    assert torch.equal(actual, expected)
+
+    # check the weights
+    # FIXME
+    # assert not torch.equal(model_signaled.layer.weight, model_total.layer.weight)
```

Comment:
I added this assertion but it fails. We should look into it

Reply:
Yes, this is normal. But I still added a FIXME for on_last_batch.